### PR TITLE
move datasources to init container.

### PR DESCRIPTION
**Pull request by xuxinkun**
To close pr https://github.com/helm/charts/pull/8637.
Datasources configmap must be collected before grafana startup. So, datasources container part is moved from `containers` to `init-containers`. And a new enviroment `METHOD=LIST` is added for datasource init container.
BTW, a new feature supporting `METHOD` is added for k8s-sidecar. If env METHOD is LIST, the program will just list the configmaps needed and then exit. To see more, pls visit https://github.com/xuxinkun/k8s-sidecar/tree/0.0.7

### CODE REVIEW

**Comment by xuxinkun** [comment]
@zanhsieh @rtluckie Pls review.

**Review by zanhsieh — APPROVED** [review]

**Comment by GreenKnight15** [comment]
@xuxinkun What is left to be done? Is it just a chart version uptick?

**Comment by xuxinkun** [comment]
@GreenKnight15 
I resolve the conflicts. And I think everything is ready for merge. Pls review.

**Comment by sameersbn** [comment]
please resolve the conflicts.
/ok-to-test

**Comment by sameersbn** [comment]
Looking at the changes it appears that `{{ template "grafana.name" . }}-sc-datasources` was already a sidecar container. I am not sure what changes this PR intends to achieve.

**Comment by GreenKnight15** [comment]
@sameersbn no in master the datasource side car is being deployed under the containers scope not under the initContainer. The data source configmaps need to be pulled in by the side car before the grafana container starts. So if the side car is in the containers scope there is a race condition. This request adds it to the initContainer scope.  @xuxinkun is that an accurate description?

**Comment by sameersbn** [comment]
got it. thanks!
/hold cancel

**Comment by xuxinkun** [comment]
@GreenKnight15  Yes, u r right.
@sameersbn I have resolved the conflicts.

**Comment by GreenKnight15** [comment]
@xuxinkun According to the failing tests I think you need to update the chart version in the Chart.yaml , follow semantic versioning.

**Comment by xuxinkun** [comment]
@GreenKnight15 I rebase my branch on master. But the test still failed. 
I check the version of chart.yaml. It is `version: 1.25.2` now. The master branch is also `version: 1.25.2`.

**Comment by xuxinkun** [comment]
@tim-beatson Oh, I see. Thank you. I will change it right now.

**Comment by xuxinkun** [comment]
@GreenKnight15 @sameersbn @tim-beatson Pls review.

**Comment by GreenKnight15** [comment]
@xuxinkun Because you changed some values in the values.yaml file and removed a container If someone tried to deploy with their own old value.yaml it would fail so, I would call that a breaking change. According to semantic versioning I would make a major revision instead of a patch. @sameersbn @tim-beatson don't you agree?

**Comment by tim-beatson** [comment]
@xuxinkun, I'm not a maintainer so I can't review, however I'm inclined to agree with @GreenKnight15, this is a breaking change so the major version should be updated.

**Comment by xuxinkun** [comment]
@GreenKnight15  Should I use `version: 1.26.1` instead？

**Comment by GreenKnight15** [comment]
@xuxinkun No that would be a minor rev, a major rev would be 2.0.0, if you only added new keys/values in the values.yaml an only added the initContainer with the new template keys it would be a major rev and be 1.26.0 because you are only adding and not removing or changing existing contracts. Follow this guide https://semver.org.

**Comment by xuxinkun** [comment]
@GreenKnight15 version has been updated to 2.0.0 . Pls review.

**Comment by GreenKnight15** [comment]
@xuxinkun I don't have permission to add the lgtm label. Which should be the last thing needed.

**Comment by maorfr** [comment]
/assign
/hold

Hey @xuxinkun,
Thanks for contributing!
There are currently some errors in the PR (the `or` conditions is one I spotted), and conflicts as well.

I have assigned myself and placed the `hold` label until these are fixed.

I will review again in a few days, hope you get a chance to fix by that time.
Thanks again, your contribution is appreciated!

**Comment by tim-beatson** [comment]
@xuxinkun, can you bump the version again pleease

**Comment by xuxinkun** [comment]
@tim-beatson Change the version again.

**Comment by tim-beatson** [comment]
@xuxinkun, please could you check the current version of the grafana chart. It's progressed significantly whilst this PR has been open. The version in this PR is older than the current released version

**Comment by stevenimle-honeywell** [comment]
I believe this broke actual data source reloading. Seems like this resolved the race condition but now datasources are not hot reloadable like the dashboards are.

**Comment by tim-beatson** [comment]
@stevenimle-honeywell, Grafana does not allow datasources to be reloaded like dashboards, that's the whole point of the PR. If datasources could be reloaded, the race condition would not exist.

**Comment by stevenimle-honeywell** [comment]
I see, then why is this still called a sidecar? It makes it sound like this is an option.

**Comment by tim-beatson** [comment]
@stevenimle-honeywell, perhaps you could suggest to @xuxinkun that he make additions to his PR to make that clearer?

**Comment by stevenimle-honeywell** [comment]
Fair enough, I appreciate the prompt reply. Since this PR is already merged however, would it not make more sense to raise this as a PR itself?

**Comment by tim-beatson** [comment]
Ah, I didn't see that this had been merged!

**Comment by xuxinkun** [comment]
@stevenimle-honeywell 
I mainly wanted to quickly implement the feature that moving the datasource sidecar container to init container at the beginning. So, I just moved the code and did not modify the `sidecar`. Maybe additional `init container` value is better. You can raise another PR to format it.

**Pull request merged**
